File: backend/src/services/menu_vocabulary.py
from __future__ import annotations

import re
import unicodedata
from typing import Final


DIET_TYPE_LABELS: Final[dict[str, str]] = {
    "regular": "常食",
    "regular_bag": "常食(袋分け)",
    "soft": "軟菜",
    "soft_mixer": "軟菜/ミキサー",
    "mixer": "ミキサー",
    "daycare": "通所",
    "staff": "職員",
    "no_fried": "禁食(揚げ物禁)",
    "tea": "お茶",
    "business": "事業",
    "diabetes": "糖尿",
    "pregnancy": "妊娠",
    "sesame_allergy": "ゴマアレルギー",
    "no_meat": "禁食(肉禁)",
    "forbidden_other": "禁食(肉卵魚禁)",
    "no_fish": "禁食(魚禁)",
    "change_1": "変更1",
    "change_2": "変更2",
    "regular_1600kcal": "常食1600kcal",
    "soft_1600kcal": "軟菜1600kcal",
    "mixer_1600kcal": "ミキサー1600kcal",
    "1600kcal": "1600kcal",
    "unknown": "不明",
}

AGGREGATED_DIET_TYPE_LABELS: Final[dict[str, str]] = {
    "regular": "常食",
    "soft": "軟菜",
    "mixer": "ミキサー",
    "forbidden": "禁食",
    "tea": "お茶",
    "business": "事業",
    "diabetes": "糖尿",
    "pregnancy": "妊娠",
    "sesame_allergy": "ゴマアレルギー",
    "change_1": "変更1",
    "change_2": "変更2",
    "1600kcal": "1600kcal",
    "unknown": "不明",
}


def normalize_diet_type(value: object) -> str | None:
    if value is None:
        return None
    text = str(value).strip()
    if not text:
        return None

    normalized = unicodedata.normalize("NFKC", text)
    compact = (
        normalized.lower()
        .replace(" ", "")
        .replace("　", "")
        .replace("_", "")
        .replace("＿", "")
    )
    compact = re.sub(r"[／/・+＋\-]", "", compact)
    compact = re.sub(r"[()（）\[\]【】]", "", compact)
    if not compact:
        return None

    base: str | None = None
    if ("袋" in compact or "bag" in compact) and (
        "regular" in compact or "常食" in compact or "通常" in compact or compact == "常"
    ):
        base = "regular_bag"
    elif "regular" in compact or "常食" in compact or "通常" in compact:
        base = "regular"
    elif "daycare" in compact or "通所" in compact:
        base = "daycare"
    elif "staff" in compact or "職員" in compact:
        base = "staff"
    elif "nofried" in compact or "揚げ物禁" in normalized or "揚物禁" in normalized:
        base = "no_fried"
    elif "tea" in compact or "お茶" in compact:
        base = "tea"
    elif "business" in compact or "事業" in compact:
        base = "business"
    elif "diabetes" in compact or "diabetic" in compact or "糖尿" in compact:
        base = "diabetes"
    elif "pregnancy" in compact or "妊娠" in compact:
        base = "pregnancy"
    elif ("ごま" in compact or "ゴマ" in normalized or "sesame" in compact) and (
        "アレル" in normalized or "allergy" in compact
    ):
        base = "sesame_allergy"
    elif (
        ("肉" in compact or "meat" in compact)
        and ("卵" in compact or "玉子" in normalized or "egg" in compact)
        and ("魚" in compact or "鯖" in normalized or "さば" in normalized or "fish" in compact)
    ) or "肉卵魚禁" in normalized:
        base = "forbidden_other"
    elif "nomeat" in compact or "nobeef" in compact or "禁食肉禁" in compact or "肉禁" in compact:
        base = "no_meat"
    elif "nofish" in compact or "禁食魚禁" in compact or "魚禁" in compact:
        base = "no_fish"
    elif "change1" in compact or "変更1" in compact:
        base = "change_1"
    elif "change2" in compact or "変更2" in compact:
        base = "change_2"
    elif compact in {"unknown", "不明"}:
        base = "unknown"
    else:
        has_soft = "soft" in compact or "軟" in compact or "やわ" in compact
        has_mixer = "mixer" in compact or "mix" in compact or "ミキサ" in compact
        if has_soft and has_mixer:
            base = "soft_mixer"
        elif has_soft:
            base = "soft"
        elif has_mixer:
            base = "mixer"

    if "1600" in compact:
        if base in {"regular", "soft", "mixer"}:
            return f"{base}_1600kcal"
        if base is None:
            return "1600kcal"
    return base or text


def format_diet_type_label(value: object) -> str:
    normalized = normalize_diet_type(value)
    if not normalized:
        return "-"
    return DIET_TYPE_LABELS.get(normalized, str(value).strip() or "-")


def bucket_diet_type_for_aggregation(value: object) -> str | None:
    normalized = normalize_diet_type(value)
    if not normalized:
        return None
    if normalized in {"regular", "regular_bag", "daycare", "staff", "regular_1600kcal"}:
        return "regular"
    if normalized in {"soft", "soft_1600kcal"}:
        return "soft"
    if normalized in {"mixer", "mixer_1600kcal"}:
        return "mixer"
    if normalized in {"no_fried", "no_meat", "no_fish", "forbidden_other"}:
        return "forbidden"
    return normalized


def format_aggregated_diet_type_label(value: object) -> str:
    normalized = bucket_diet_type_for_aggregation(value)
    if not normalized:
        return "-"
    return AGGREGATED_DIET_TYPE_LABELS.get(normalized, str(value).strip() or "-")

File: frontend/src/services/menuVocabulary.ts
const normalizeCompact = (value?: string | null) =>
  (value || "")
    .trim()
    .toLowerCase()
    .replace(/[\s　]+/g, "")
    .replace(/[＿_]/g, "")
    .replace(/[／/・+＋-]/g, "")
    .replace(/[()（）\[\]【】]/g, "");

export const DIET_TYPE_OPTIONS = [
  { value: "", label: "未選択" },
  { value: "regular", label: "常食" },
  { value: "regular_bag", label: "常食(袋分け)" },
  { value: "soft", label: "軟菜" },
  { value: "soft_mixer", label: "軟菜/ミキサー" },
  { value: "mixer", label: "ミキサー" },
  { value: "daycare", label: "通所" },
  { value: "staff", label: "職員" },
  { value: "tea", label: "お茶" },
  { value: "business", label: "事業" },
  { value: "diabetes", label: "糖尿" },
  { value: "pregnancy", label: "妊娠" },
  { value: "sesame_allergy", label: "ゴマアレルギー" },
  { value: "no_fried", label: "禁食(揚げ物禁)" },
  { value: "no_meat", label: "禁食(肉禁)" },
  { value: "forbidden_other", label: "禁食(肉卵魚禁)" },
  { value: "no_fish", label: "禁食(魚禁)" },
  { value: "change_1", label: "変更1" },
  { value: "change_2", label: "変更2" },
  { value: "regular_1600kcal", label: "常食1600kcal" },
  { value: "soft_1600kcal", label: "軟菜1600kcal" },
  { value: "mixer_1600kcal", label: "ミキサー1600kcal" },
  { value: "1600kcal", label: "1600kcal" },
];

const DIET_TYPE_LABELS = Object.fromEntries(
  DIET_TYPE_OPTIONS.filter((option) => option.value).map((option) => [option.value, option.label])
) as Record<string, string>;

export const normalizeDietTypeValue = (value?: string | null) => {
  const raw = (value || "").trim();
  if (!raw) return "";
  const compact = normalizeCompact(raw);
  if (!compact) return "";
  if ((compact.includes("袋") || compact.includes("bag")) && (compact.includes("regular") || compact.includes("常食") || compact.includes("通常") || compact === "常")) {
    return "regular_bag";
  }
  if (compact.includes("regular") || compact.includes("常食") || compact.includes("通常")) return "regular";
  if (compact.includes("daycare") || compact.includes("通所")) return "daycare";
  if (compact.includes("staff") || compact.includes("職員")) return "staff";
  if (compact.includes("tea") || compact.includes("お茶")) return "tea";
  if (compact.includes("business") || compact.includes("事業")) return "business";
  if (compact.includes("diabetes") || compact.includes("diabetic") || compact.includes("糖尿")) return "diabetes";
  if (compact.includes("pregnancy") || compact.includes("妊娠")) return "pregnancy";
  if ((compact.includes("ごま") || compact.includes("sesame")) && (raw.includes("アレル") || compact.includes("allergy"))) {
    return "sesame_allergy";
  }
  if (compact.includes("揚げ物禁") || compact.includes("揚物禁") || compact.includes("nofried") || compact.includes("friedfree")) return "no_fried";
  if (
    (compact.includes("肉") || compact.includes("meat")) &&
    (compact.includes("卵") || compact.includes("玉子") || compact.includes("egg")) &&
    (compact.includes("魚") || compact.includes("鯖") || compact.includes("さば") || compact.includes("fish"))
  ) {
    return "forbidden_other";
  }
  if (compact.includes("nomeat") || compact.includes("nobeef") || compact.includes("禁食肉禁") || compact.includes("肉禁")) return "no_meat";
  if (compact.includes("nofish") || compact.includes("禁食魚禁") || compact.includes("魚禁")) return "no_fish";
  if (compact.includes("change1") || compact.includes("変更1")) return "change_1";
  if (compact.includes("change2") || compact.includes("変更2")) return "change_2";
  const hasSoft = compact.includes("soft") || compact.includes("軟");
  const hasMixer = compact.includes("mixer") || compact.includes("mix") || compact.includes("ミキサ");
  if (hasSoft && hasMixer) return "soft_mixer";
  if (hasSoft) return "soft";
  if (hasMixer) return "mixer";
  if (compact.includes("1600")) {
    if (compact.includes("常食") || compact.includes("regular")) return "regular_1600kcal";
    if (compact.includes("軟") || compact.includes("soft")) return "soft_1600kcal";
    if (compact.includes("ミキサ") || compact.includes("mixer")) return "mixer_1600kcal";
    return "1600kcal";
  }
  return raw;
};

export const formatDietTypeLabel = (value?: string | null) => {
  const normalized = normalizeDietTypeValue(value);
  if (!normalized) return "-";
  return DIET_TYPE_LABELS[normalized] || value || "-";
};
